django/db/migrations/operations/models.py

```diff
@@ -407,21 +407,39 @@ def database_forwards(self, app_label, schema_editor, from_state, to_state):
                     or not new_field.remote_field.through._meta.auto_created
                 ):
                     continue
-                # Rename the M2M table that's based on this model's name.
+
                 old_m2m_model = old_field.remote_field.through
                 new_m2m_model = new_field.remote_field.through
-                schema_editor.alter_db_table(
-                    new_m2m_model,
-                    old_m2m_model._meta.db_table,
-                    new_m2m_model._meta.db_table,
-                )
-                # Rename the column in the M2M table that's based on this
-                # model's name.
-                schema_editor.alter_field(
-                    new_m2m_model,
-                    old_m2m_model._meta.get_field(old_model._meta.model_name),
-                    new_m2m_model._meta.get_field(new_model._meta.model_name),
-                )
+
+                # Rename both columns and the table name of the M2M table
+                # in case of same model
+                if (
+                    old_field.related_model._meta.model_name
+                    == old_model._meta.model_name
+                    or new_field.related_model._meta.model_name
+                    == new_model._meta.model_name
+                ):
+                    schema_editor.alter_field(
+                        new_model,
+                        old_field,
+                        new_field,
+                        strict=False,
+                        force=True,
+                    )
+                else:
+                    # Rename the M2M table that's based on this model's name.
+                    schema_editor.alter_db_table(
+                        new_m2m_model,
+                        old_m2m_model._meta.db_table,
+                        new_m2m_model._meta.db_table,
+                    )
+                    # Rename the column in the M2M table that's based on this
+                    # model's name.
+                    schema_editor.alter_field(
+                        new_m2m_model,
+                        old_m2m_model._meta.get_field(old_model._meta.model_name),
+                        new_m2m_model._meta.get_field(new_model._meta.model_name),
+                    )
```

Comment:
What I meant by that is that this whole thing can be simplified to a single `alter_field(force=True)` call

```python
schema_editor.alter_field(
    new_model,
    old_field,
    new_field,
    strict=False,
    force=True,
)
```

No need for `if`/`else` of anything as `_alter_many_to_many` already takes care of renaming table systematically and its `alter_field` calls are gated by `_field_should_be_altered` checks to avoid unnecessary alterations.